test/y2add_on/clients/add-on_auto_test.rb

```diff
@@ -71,19 +290,88 @@
           ]
         end
 
+
         before do
           allow(Yast::AddOnProduct).to receive(:add_on_products).and_return(add_on_products)
           allow(Yast::Pkg).to receive(:SourceEditSet)
           allow(Yast::Pkg).to receive(:SourceCreate).and_return(1)
           allow(Yast::Pkg).to receive(:SourceEditGet).and_return(repos)
         end
 
+        # FIXME: improve that WIP scenarios/contexts
+        context "and product creation fails" do
+          before do
+            allow(Yast::Pkg).to receive(:SourceCreate).and_return(-1)
+          end
+
+          context "ask_on_error=true" do
+            let(:ask_on_error) { true }
+
+            it "ask to make it available" do
+              expect(Yast::Popup).to receive(:ContinueCancel)
+
+              subject.run
+            end
+          end
+
+
+          context "ask_on_error=false" do
+            before do
+              allow(Yast::Popup).to receive(:ContinueCancel).and_return(false)
+            end
+
+            let(:ask_on_error) { false }
+
+            it "report error" do
+              expect(Yast::Report).to receive(:Error)
+
+              subject.run
+            end
+          end
+        end
+
         it "stores repos according to information given" do
           expect(Yast::Pkg).to receive(:SourceEditSet).with(repos_to_store)
 
-          subject.main
+          subject.run
         end
       end
     end
+
+    context "when 'func' is 'Read'" do
+      let(:func) { "Read" }
+
+      context "and cannot lock package" do
+        before do
+          allow(Yast::PackageLock).to receive(:Check).and_return(false)
+        end
+
+        it "returns false" do
+          expect(subject.run).to be_falsey
+        end
+      end
+
+      context "and can lock package" do
+        before do
+          allow(Yast::PackageLock).to receive(:Check).and_return(true)
+          allow(Yast::Pkg).to receive(:SourceStartManager)
+        end
+
+        it "reads add-ons configuration from the current system" do
+          expect(subject).to receive(:ReadFromSystem)
+
+          subject.run
+        end
+      end
+    end
+
+    # FIXME: delete it? probably already tested in AutoClient
```

Comment:
yes, whole dispathing is tested in auto client. In general I have rule for unit tests that I assume that other units works and I test only my unit.
https://github.com/yast/yast-yast2/blob/0d1f6001e99734c8da163d78c04bf53b029ba8c4/library/general/test/auto_client_test.rb

Reply:
Yes, you should only test the code implemented in your module. Testing the rest is out of scope and makes it fragile if the external code changes.